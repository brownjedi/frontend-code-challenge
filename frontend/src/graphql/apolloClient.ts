/* eslint-disable graphql/template-strings */
import { PokemonList } from '@/types'
import { ApolloClient, createHttpLink, InMemoryCache } from '@apollo/client/core'
import { resolvers } from './resolvers'
import { typeDefs } from './typeDefs'
import produce from 'immer'

// HTTP connection to the API
const httpLink = createHttpLink({
  // You should use an absolute URL here
  uri: 'http://localhost:4000/graphql',
})

// Cache implementation
const cache = new InMemoryCache({
  typePolicies: {
    Query: {
      fields: {
        pokemons: {
          keyArgs: ['query', ['search', 'filter', ['type', 'isFavorite']]],
          merge(existing: PokemonList, incoming: PokemonList, { args }) {
            if (!existing) {
              return incoming
            }

            const offset = args?.query?.offset || 0
            const nextState = produce(existing, draftState => {
              for (let i = 0; i < incoming.edges.length; ++i) {
                draftState.edges[offset + i] = incoming.edges[i]
              }
            })

            return nextState
          },
        },
      },
    },
  },
})

// Create the apollo client
export const apolloClient = new ApolloClient({
  link: httpLink,
  cache,
  typeDefs,
  resolvers,
  defaultOptions: {
    watchQuery: {
      refetchWritePolicy: 'overwrite',
    },
  },
})
