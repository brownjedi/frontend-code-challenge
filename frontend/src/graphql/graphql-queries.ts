import { gql } from '@apollo/client/core'

const POKEMON_OVERVIEW_FRAGMENT = gql`
  fragment PokemonOverviewFields on Pokemon {
    id
    name
    image
    sprites {
      normal
      animated
    }
    classification
    types
    isFavorite
  }
`

export const POKEMONS_QUERY = gql`
  ${POKEMON_OVERVIEW_FRAGMENT}
  query pokemons($query: PokemonsQueryInput! = { limit: 30, offset: 0 }) {
    pokemons(query: $query) {
      limit
      offset
      count
      edges {
        ...PokemonOverviewFields
      }
    }
  }
`

export const POKEMON_TYPES_QUERY = gql`
  query pokemonTypes {
    pokemonTypes
  }
`

export const POKEMON_BY_NAME_QUERY = gql`
  ${POKEMON_OVERVIEW_FRAGMENT}
  query pokemon($name: String!) {
    pokemon: pokemonByName(name: $name) {
      ...PokemonOverviewFields
      number
      maxCP
      maxHP
      fleeRate
      resistant
      sound
      types
      weaknesses
      rgbaster @client {
        backgroundColor
        textColor
      }
      attacks {
        special {
          name
          type
          damage
        }
        fast {
          name
          type
          damage
        }
      }
      evolutionRequirements {
        name
        amount
      }
      evolutions {
        id
        name
      }
      height {
        minimum
        maximum
      }
      weight {
        minimum
        maximum
      }
    }
  }
`

export const FAVORITE_POKEMON_MUTATION = gql`
  mutation favoritePokemon($id: ID!) {
    pokemon: favoritePokemon(id: $id) {
      id
      isFavorite
    }
  }
`

export const UNFAVORITE_POKEMON_MUTATION = gql`
  mutation unFavoritePokemon($id: ID!) {
    pokemon: unFavoritePokemon(id: $id) {
      id
      isFavorite
    }
  }
`
