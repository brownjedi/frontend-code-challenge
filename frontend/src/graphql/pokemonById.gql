query pokemon($id: ID!) {
  pokemon: pokemonById(id: $id) {
    id
    name
    number
    classification
    image
    isFavorite
    maxCP
    maxHP
    fleeRate
    resistant
    sound
    types
    weaknesses
    attacks {
      special {
        name
        type
        damage
      }
      fast {
        name
        type
        damage
      }
    }
    evolutionRequirements {
      name
      amount
    }
    evolutions {
      id
      name
    }
    height {
      minimum
      maximum
    }
    weight {
      minimum
      maximum
    }
  }
}
