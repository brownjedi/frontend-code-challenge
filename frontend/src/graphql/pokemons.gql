query pokemons {
  pokemons(query: {}) {
    limit
    offset
    edges {
      id
      name
    }
  }
}
