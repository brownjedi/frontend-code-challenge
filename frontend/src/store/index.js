import Vue from 'vue'
import Vuex from 'vuex'

// Load all modules.
function loadModules() {
  const context = require.context('./modules', false, /([a-z_]+)\.js$/i)

  const modules = context
    .keys()
    .map(key => ({ key, name: key.match(/([a-z_]+)\.js$/i)[1] }))
    .reduce(
      (modules, { key, name }) => ({
        ...modules,
        [name]: context(key).default,
      }),
      {}
    )

  return { context, modules }
}

const { modules } = loadModules()

Vue.use(Vuex)

export default new Vuex.Store({
  strict: process.env.NODE_ENV !== 'production',
  modules,
})
