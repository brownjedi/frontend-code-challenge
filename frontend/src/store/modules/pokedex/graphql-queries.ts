import gql from 'graphql-tag'

const POKEMON_OVERVIEW_FRAGMENT = gql`
  fragment PokemonOverviewFields on Pokemon {
    id
    name
    image
    sprites {
      normal
      animated
    }
    classification
    types
    isFavorite
  }
`

export const POKEMONS_QUERY = gql`
  ${POKEMON_OVERVIEW_FRAGMENT}
  query pokemons {
    pokemons(query: { limit: 100 }) {
      limit
      offset
      edges {
        ...PokemonOverviewFields
      }
    }
  }
`

export const POKEMON_BY_NAME_QUERY = gql`
  ${POKEMON_OVERVIEW_FRAGMENT}
  query pokemon($name: String!) {
    pokemon: pokemonByName(name: $name) {
      ...PokemonOverviewFields
      number
      maxCP
      maxHP
      fleeRate
      resistant
      sound
      types
      weaknesses
      attacks {
        special {
          name
          type
          damage
        }
        fast {
          name
          type
          damage
        }
      }
      evolutionRequirements {
        name
        amount
      }
      evolutions {
        id
        name
      }
      height {
        minimum
        maximum
      }
      weight {
        minimum
        maximum
      }
    }
  }
`
