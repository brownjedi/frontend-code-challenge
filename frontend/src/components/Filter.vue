<template>
  <div :class="styles.container">
    <transition name="fade">
      <!-- Loading -->
      <div v-if="loading" :class="styles.loading">
        <FilterButtonSkeleton v-for="index in 18" :key="index" :class="styles.button" />
      </div>

      <!-- Error -->
      <div v-else-if="error" :class="styles.error">An error occured</div>

      <!-- Result -->
      <div v-else-if="types" :class="styles.result">
        <FilterButton
          type="button"
          :class="styles.button"
          @click="(val, clr) => onClick('', clr)"
          :active="selectedType === ''"
          text="All"
        />
        <FilterButton
          type="button"
          :class="styles.button"
          @click="(val, clr) => onClick(val, clr)"
          :active="selectedType === type"
          v-for="type in types"
          :key="type"
          :text="type"
        />
      </div>

      <!-- No result -->
      <div v-else :class="styles['no-result']">No result :(</div>
    </transition>
  </div>
</template>

<script lang="ts">
import { computed, defineComponent, useCssModule } from 'vue'
import { usePokemonTypesQuery } from '@/composables/pokemonsTypesQuery'
import FilterButton from './FilterButton.vue'
import FilterButtonSkeleton from './FilterButtonSkeleton.vue'
import { useStore } from '@/store'
import { PokedexMutations } from '@/store/modules/pokedex/mutations'

export default defineComponent({
  components: { FilterButton, FilterButtonSkeleton },
  setup() {
    const styles = useCssModule()
    const store = useStore()
    const selectedType = computed(() => store.state.pokedex.query.filter.type)
    const { types, loading, error } = usePokemonTypesQuery()

    function onClick(type: string, color: string) {
      store.commit(PokedexMutations.SET_TYPE, { type })
      store.commit(PokedexMutations.SET_COLOR, { color })
    }

    return { styles, types, loading, error, selectedType, onClick }
  },
})
</script>

<style lang="scss" module>
.container {
  height: 100vh;
  box-shadow: 0 0 40px var(--box-shadow-color);
  position: sticky;
  top: 0;
  flex: 0 0 auto;
  padding: 0 0.625rem;
  overflow: auto;
}

.loading,
.result,
.error,
.no-result {
  display: flex;
  flex-direction: column;
}

.button {
  flex: 0 0 auto;
  width: 6.75rem;
}

.button + .button {
  margin-top: 0.625rem;
}

.button:first-child {
  margin-top: 0.625rem;
}

.button:last-child {
  margin-bottom: 0.625rem;
}
</style>
