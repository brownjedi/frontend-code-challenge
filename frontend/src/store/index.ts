import { createStore, createLogger } from 'vuex'
import { RootState } from './types'

import { store as pokedex, PokedexStore } from './modules/pokedex'

export type Store = PokedexStore<Pick<RootState, 'pokedex'>>

export const store = createStore<RootState>({
  plugins: [createLogger()],
  modules: {
    pokedex,
  },
})

export function useStore(): Store {
  return store as Store
}
