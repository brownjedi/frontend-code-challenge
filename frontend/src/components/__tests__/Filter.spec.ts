import { mount, flushPromises } from '@vue/test-utils'
import { createMockClient, MockApolloClient } from 'mock-apollo-client'
import { provideApolloClient } from '@vue/apollo-composable'
import Filter from '@/components/Filter.vue'
import { POKEMON_TYPES_QUERY } from '@/graphql/graphql-queries'
import { store } from '@/store'

describe('components/Filter.vue', () => {
  let requestHandler: jest.Mock
  let mockClient: MockApolloClient

  beforeEach(() => {
    mockClient = createMockClient()
    requestHandler = jest.fn()
    mockClient.setRequestHandler(POKEMON_TYPES_QUERY, requestHandler)
    provideApolloClient(mockClient)
  })

  describe('when query loading', () => {
    it('has the expected html structure', async () => {
      const wrapper = mount(Filter)
      expect(wrapper.element).toMatchSnapshot()
    })
  })

  describe('when query fails', () => {
    it('shows error message', async () => {
      requestHandler.mockRejectedValueOnce(new Error('Graphql query failed'))

      const wrapper = mount(Filter)

      expect(requestHandler).toHaveBeenCalledTimes(1)

      // wait for apollo results
      await flushPromises()

      // ensure that result container exists
      expect(wrapper.find('.error').exists()).toBe(true)
      expect(wrapper.find('.error').text()).toBe('An error occured')

      // ensure we have zero buttons when errors
      expect(wrapper.findAll('button').length).toBe(0)
    })
  })

  describe('when query succeeds', () => {
    it('shows no result message when no data', async () => {
      requestHandler.mockResolvedValueOnce({
        data: { pokemonTypes: null },
      })

      const wrapper = mount(Filter)

      expect(requestHandler).toHaveBeenCalledTimes(1)

      // wait for apollo results
      await flushPromises()

      // ensure that result container exists
      expect(wrapper.find('.no-result').exists()).toBe(true)
      expect(wrapper.find('.no-result').text()).toBe('No result :(')

      // ensure we have zero buttons when errors
      expect(wrapper.findAll('button').length).toBe(0)
    })

    it('renders results correctly when data is present', async () => {
      const mapping: Record<string, string> = { Fire: '#FF0000', Ice: '#0000FF', Grass: '#00FF00', Dragon: '$EDEDED' }
      const pokemonTypes = Object.keys(mapping)
      requestHandler.mockResolvedValueOnce({
        data: { pokemonTypes },
      })

      // document setup
      document.body.innerHTML = `
      <div>
        <h1>Non Vue app</h1>
        <div id="app"></div>
      </div>
    `
      document.documentElement.style.setProperty(`--main-text-color`, '#000000')
      Object.keys(mapping).forEach(x => {
        document.documentElement.style.setProperty(`--${x.toLowerCase()}-text-color`, mapping[x])
      })

      const wrapper = mount(Filter, { attachTo: '#app', global: { plugins: [store] } })

      expect(requestHandler).toHaveBeenCalledTimes(1)

      // wait for apollo results
      await flushPromises()

      // ensure that result container exists
      expect(wrapper.find('.result').exists()).toBe(true)

      const buttons = wrapper.findAll('button')

      // ensure we have (n + 1) buttons. The first button is All button
      expect(buttons.length).toBe(pokemonTypes.length + 1)

      // first button should be the All button and selected by default
      expect(buttons[0].classes()).toContain('active')
      expect(buttons[0].text()).toBe('All')
      await buttons[0].trigger('click')
      expect(store.state.pokedex.query.filter.type).toStrictEqual('')
      expect(store.state.pokedex.ui.color).toStrictEqual('#000000')

      // check the values of other buttons and the vuex store data
      for (let i = 0; i < pokemonTypes.length; i++) {
        expect(buttons[i + 1].text()).toBe(pokemonTypes[i])
        await buttons[i + 1].trigger('click')
        expect(store.state.pokedex.query.filter.type).toStrictEqual(pokemonTypes[i])
        expect(store.state.pokedex.ui.color).toStrictEqual(mapping[pokemonTypes[i]])
      }
    })
  })
})
