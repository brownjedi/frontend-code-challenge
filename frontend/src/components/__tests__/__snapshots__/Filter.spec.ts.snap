// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/Filter.vue when query loading has the expected html structure 1`] = `
<div
  class="container"
>
  <transition-stub>
    <div
      class="loading"
    >
      
      <div
        class="button"
      >
        <div
          class="skeleton"
        />
      </div>
      <div
        class="button"
      >
        <div
          class="skeleton"
        />
      </div>
      <div
        class="button"
      >
        <div
          class="skeleton"
        />
      </div>
      <div
        class="button"
      >
        <div
          class="skeleton"
        />
      </div>
      <div
        class="button"
      >
        <div
          class="skeleton"
        />
      </div>
      <div
        class="button"
      >
        <div
          class="skeleton"
        />
      </div>
      <div
        class="button"
      >
        <div
          class="skeleton"
        />
      </div>
      <div
        class="button"
      >
        <div
          class="skeleton"
        />
      </div>
      <div
        class="button"
      >
        <div
          class="skeleton"
        />
      </div>
      <div
        class="button"
      >
        <div
          class="skeleton"
        />
      </div>
      <div
        class="button"
      >
        <div
          class="skeleton"
        />
      </div>
      <div
        class="button"
      >
        <div
          class="skeleton"
        />
      </div>
      <div
        class="button"
      >
        <div
          class="skeleton"
        />
      </div>
      <div
        class="button"
      >
        <div
          class="skeleton"
        />
      </div>
      <div
        class="button"
      >
        <div
          class="skeleton"
        />
      </div>
      <div
        class="button"
      >
        <div
          class="skeleton"
        />
      </div>
      <div
        class="button"
      >
        <div
          class="skeleton"
        />
      </div>
      <div
        class="button"
      >
        <div
          class="skeleton"
        />
      </div>
      
    </div>
  </transition-stub>
</div>
`;
